File: packages/arcis-python/arcis/middleware/token_budget.py
"""
Token-budget protection middleware for Python.

Caps per-key token spend over a sliding window — meant for routes that proxy
LLM calls, where a tight 100-req/min rate limit isn't enough because a single
50KB prompt costs the same as 1000 small requests.

Mirrors the Node implementation at
``packages/arcis-node/src/middleware/token-budget.ts``.

Usage with Flask:

.. code-block:: python

    from flask import Flask, request, jsonify
    from arcis.middleware.token_budget import TokenBudget, TokenBudgetExceeded

    app = Flask(__name__)
    guard = TokenBudget(max_tokens=100_000, window_seconds=3600,
                        max_request_tokens=5_000)

    @app.post("/chat")
    def chat():
        try:
            guard.check(request)
        except TokenBudgetExceeded as e:
            return jsonify(e.to_dict()), e.status_code
        # ...

Or as a FastAPI dependency:

.. code-block:: python

    from fastapi import FastAPI, Request, Depends, HTTPException
    from arcis.middleware.token_budget import TokenBudget, TokenBudgetExceeded

    guard = TokenBudget(max_tokens=100_000)

    async def enforce_budget(request: Request):
        try:
            await guard.acheck(request)
        except TokenBudgetExceeded as e:
            raise HTTPException(status_code=e.status_code, detail=e.message)

    @app.post("/chat", dependencies=[Depends(enforce_budget)])
    async def chat(): ...
"""

from __future__ import annotations

import json
import threading
import time
from dataclasses import dataclass
from typing import Any, Callable, Dict, Optional


# ─── Defaults ──────────────────────────────────────────────────────────────

_DEFAULT_MAX_TOKENS = 100_000
_DEFAULT_WINDOW_SECONDS = 60 * 60  # 1 hour
_DEFAULT_MESSAGE = "Token budget exceeded for this window."
_DEFAULT_MESSAGE_OVERSIZE = "Request exceeds the per-request token limit."


@dataclass
class _Entry:
    used: int
    reset_time: float


class TokenBudgetExceeded(Exception):
    """Raised when a request would push a key past its token budget.

    Carries enough information for the caller to construct a 429 (or 413
    when ``oversize=True``) response with proper headers.
    """

    def __init__(
        self,
        message: str,
        *,
        status_code: int,
        used: int,
        max_tokens: int,
        request_tokens: int,
        retry_after_seconds: int,
        oversize: bool = False,
    ):
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.used = used
        self.max_tokens = max_tokens
        self.request_tokens = request_tokens
        self.retry_after_seconds = retry_after_seconds
        self.oversize = oversize

    def to_dict(self) -> Dict[str, Any]:
        body: Dict[str, Any] = {"error": self.message}
        if self.oversize:
            body["requestTokens"] = self.request_tokens
            body["maxRequestTokens"] = self.max_tokens
        else:
            body["used"] = self.used
            body["maxTokens"] = self.max_tokens
            body["retryAfter"] = self.retry_after_seconds
        return body


def _default_key_generator(request: Any) -> str:
    """Default key: client IP, fall back to ``unknown``.

    Reads the IP from the most common shapes (Flask ``request.remote_addr``,
    Starlette/FastAPI ``request.client.host``, raw mappings) without
    depending on any web framework being installed.
    """
    if hasattr(request, "remote_addr"):
        ip = getattr(request, "remote_addr", None)
        if ip:
            return str(ip)
    client = getattr(request, "client", None)
    if client is not None and getattr(client, "host", None):
        return str(client.host)
    if isinstance(request, dict):
        ip = request.get("ip") or request.get("remote_addr")
        if ip:
            return str(ip)
    return "unknown"


def _default_estimate_tokens(request: Any) -> int:
    """Approximate OpenAI's "1 token ≈ 4 characters" rule.

    Reads body / json / query off whatever shape the request happens to be,
    falling back to ``getattr`` so it works with Flask, FastAPI, Django, and
    plain mappings without import-time framework coupling.
    """

    def _byte_length(obj: Any) -> int:
        if obj is None:
            return 0
        if isinstance(obj, (bytes, bytearray)):
            return len(obj)
        if isinstance(obj, str):
            return len(obj.encode("utf-8"))
        try:
            return len(json.dumps(obj, default=str).encode("utf-8"))
        except (TypeError, ValueError):
            return 0

    bytes_total = 0
    # Flask: request.get_json(silent=True) / request.data / request.form
    if hasattr(request, "get_json"):
        try:
            json_body = request.get_json(silent=True)  # type: ignore[call-arg]
        except (TypeError, ValueError):
            json_body = None
        if json_body is not None:
            bytes_total += _byte_length(json_body)
    if hasattr(request, "data"):
        bytes_total += _byte_length(getattr(request, "data", None))
    # FastAPI/Starlette: request.body would be async — use cached attrs only
    cached_body = getattr(request, "_body", None)
    if cached_body is not None:
        bytes_total += _byte_length(cached_body)
    # Plain Python mapping (tests, custom adapters)
    if isinstance(request, dict):
        bytes_total += _byte_length(request.get("body"))
        bytes_total += _byte_length(request.get("query"))
    # Query string
    if hasattr(request, "args"):
        try:
            bytes_total += _byte_length(dict(request.args.items()))
        except (TypeError, ValueError):
            pass
    if hasattr(request, "query_params"):
        try:
            bytes_total += _byte_length(dict(request.query_params))
        except (TypeError, ValueError):
            pass
    return max(0, -(-bytes_total // 4))  # ceil division by 4


# ─── Public API ────────────────────────────────────────────────────────────


class TokenBudget:
    """Per-key sliding-window token-budget enforcer.

    Thread-safe in-memory store — fine for single-process apps. For
    multi-instance deployments wire a custom store via ``key_generator`` that
    proxies to Redis (Phase B).
    """

    def __init__(
        self,
        max_tokens: int = _DEFAULT_MAX_TOKENS,
        window_seconds: float = _DEFAULT_WINDOW_SECONDS,
        max_request_tokens: Optional[int] = None,
        key_generator: Optional[Callable[[Any], str]] = None,
        estimate_tokens: Optional[Callable[[Any], int]] = None,
        status_code: int = 429,
        status_code_oversize: int = 413,
        message: str = _DEFAULT_MESSAGE,
        message_oversize: str = _DEFAULT_MESSAGE_OVERSIZE,
        skip: Optional[Callable[[Any], bool]] = None,
    ):
        self.max_tokens = max_tokens
        self.window_seconds = window_seconds
        self.max_request_tokens = max_request_tokens
        self.key_generator = key_generator or _default_key_generator
        self.estimate_tokens = estimate_tokens or _default_estimate_tokens
        self.status_code = status_code
        self.status_code_oversize = status_code_oversize
        self.message = message
        self.message_oversize = message_oversize
        self.skip = skip
        self._store: Dict[str, _Entry] = {}
        self._lock = threading.Lock()

    # ── core ────────────────────────────────────────────────────────────

    def check(self, request: Any) -> Dict[str, Any]:
        """Charge this request against the budget. Raise on violations.

        Raises:
            TokenBudgetExceeded: When the request alone exceeds the
                per-request cap, OR when the projected window total would
                exceed ``max_tokens``.

        Returns:
            Dict of headers to apply to the eventual response — caller is
            expected to forward these to the framework's response object.
        """
        if self.skip is not None and self.skip(request):
            return {}

        try:
            estimated = self.estimate_tokens(request)
        except Exception:
            estimated = 0
        if not isinstance(estimated, int) or estimated < 0:
            estimated = 0

        # Per-request cap (rejected before charging the budget)
        if (
            self.max_request_tokens is not None
            and estimated > self.max_request_tokens
        ):
            raise TokenBudgetExceeded(
                self.message_oversize,
                status_code=self.status_code_oversize,
                used=0,
                max_tokens=self.max_request_tokens,
                request_tokens=estimated,
                retry_after_seconds=0,
                oversize=True,
            )

        key = self.key_generator(request)
        now = time.monotonic()

        with self._lock:
            entry = self._store.get(key)
            if entry is None or entry.reset_time < now:
                entry = _Entry(used=0, reset_time=now + self.window_seconds)
                self._store[key] = entry

            projected = entry.used + estimated
            reset_seconds = max(0, int(entry.reset_time - now + 0.999))

            if projected > self.max_tokens:
                raise TokenBudgetExceeded(
                    self.message,
                    status_code=self.status_code,
                    used=entry.used,
                    max_tokens=self.max_tokens,
                    request_tokens=estimated,
                    retry_after_seconds=reset_seconds,
                )

            entry.used = projected
            remaining = max(0, self.max_tokens - entry.used)

            return {
                "X-Token-Budget-Limit": str(self.max_tokens),
                "X-Token-Budget-Used": str(entry.used),
                "X-Token-Budget-Remaining": str(remaining),
                "X-Token-Budget-Reset": str(reset_seconds),
                "X-Token-Budget-Request-Cost": str(estimated),
            }

    async def acheck(self, request: Any) -> Dict[str, Any]:
        """Async wrapper around :meth:`check` for FastAPI/Starlette."""
        return self.check(request)

    # ── inspection ──────────────────────────────────────────────────────

    def inspect(self, key: str) -> Optional[Dict[str, float]]:
        """Read-only view of a key's current budget state. Returns None if
        the key has never been charged (or its window has expired)."""
        with self._lock:
            entry = self._store.get(key)
            if entry is None:
                return None
            return {"used": entry.used, "reset_time": entry.reset_time}

    def reset(self, key: Optional[str] = None) -> None:
        """Clear a single key's budget, or every key if ``key`` is None.
        Mainly for tests."""
        with self._lock:
            if key is None:
                self._store.clear()
            else:
                self._store.pop(key, None)


def token_budget(**kwargs: Any) -> TokenBudget:
    """Convenience factory mirroring Node's ``tokenBudget(opts)`` API."""
    return TokenBudget(**kwargs)


__all__ = [
    "TokenBudget",
    "TokenBudgetExceeded",
    "token_budget",
]

File: packages/arcis-node/src/middleware/token-budget.ts
/**
 * @module @arcis/node/middleware/token-budget
 *
 * Token-budget protection middleware. Caps per-key token spend over a
 * sliding window — meant for routes that proxy LLM calls, where a tight
 * 100-req/min rate limit isn't enough because a single 50KB prompt costs
 * the same as 1000 small requests.
 *
 * @example
 * import express from 'express';
 * import { tokenBudget } from '@arcis/node';
 *
 * const guard = tokenBudget({
 *   maxTokens: 100_000,            // 100K tokens per window
 *   windowMs: 60 * 60 * 1000,      // 1-hour window
 *   maxRequestTokens: 5_000,       // reject any single request over this
 *   keyGenerator: (req) => req.user?.id ?? req.ip,
 * });
 *
 * app.post('/chat', guard, chatHandler);
 *
 * The default token estimator is `Math.ceil(bytesOf(body+query) / 4)` —
 * close to OpenAI's "1 token ≈ 4 English characters" rule. Override via
 * `estimateTokens` for accurate counting (tiktoken, etc.).
 */

import type { Request, RequestHandler } from 'express';

export interface TokenBudgetOptions {
  /** Max tokens a single key can spend in one window. Default: 100,000. */
  maxTokens?: number;
  /** Window length in milliseconds. Default: 60 * 60 * 1000 (1 hour). */
  windowMs?: number;
  /**
   * Max tokens a single request may consume. Requests over this size are
   * rejected with 413 BEFORE counting against the window budget. Default:
   * unset (no per-request cap).
   */
  maxRequestTokens?: number;
  /**
   * Function that returns the budget key for a request. Default: client IP
   * (req.ip), falling back to `unknown` when unresolvable.
   */
  keyGenerator?: (req: Request) => string;
  /**
   * Function that estimates the number of tokens a request will consume.
   * Default: `Math.ceil((req.body + req.query stringified bytes) / 4)`,
   * which approximates OpenAI's "1 token ≈ 4 characters" rule. Override
   * with tiktoken/accurate counting when you need precision.
   */
  estimateTokens?: (req: Request) => number;
  /** Status code for budget-exceeded responses. Default: 429. */
  statusCode?: number;
  /** Status code for oversize-request rejections. Default: 413. */
  statusCodeOversize?: number;
  /** Error message when budget is exceeded. */
  message?: string;
  /** Error message when a single request exceeds maxRequestTokens. */
  messageOversize?: string;
  /** Skip budget enforcement for certain requests. */
  skip?: (req: Request) => boolean;
}

export interface TokenBudgetMiddleware extends RequestHandler {
  /** Release internal cleanup resources. */
  close: () => void;
  /** Inspect current usage for a key (read-only; for tests/telemetry). */
  inspect: (key: string) => { used: number; resetTime: number } | null;
}

interface Entry {
  used: number;
  resetTime: number;
}

const DEFAULT_MAX_TOKENS = 100_000;
const DEFAULT_WINDOW_MS = 60 * 60 * 1000;

function defaultKeyGenerator(req: Request): string {
  const ip = req.ip ?? req.socket?.remoteAddress;
  return ip ?? 'unknown';
}

function defaultEstimateTokens(req: Request): number {
  // Approximate OpenAI's "1 token ≈ 4 characters" rule using byte length
  // of stringified body + query. Fast, side-effect-free, no model call.
  // Returns 0 for missing payloads so safe routes don't get charged.
  let bytes = 0;
  if (req.body !== undefined && req.body !== null) {
    try {
      bytes += Buffer.byteLength(
        typeof req.body === 'string' ? req.body : JSON.stringify(req.body),
        'utf8',
      );
    } catch {
      // Circular structure or similar — treat as 0 rather than throw.
    }
  }
  if (req.query && Object.keys(req.query).length > 0) {
    try {
      bytes += Buffer.byteLength(JSON.stringify(req.query), 'utf8');
    } catch {
      // ignore
    }
  }
  return Math.ceil(bytes / 4);
}

/**
 * Build a token-budget middleware. See module-level JSDoc for usage.
 */
export function tokenBudget(options: TokenBudgetOptions = {}): TokenBudgetMiddleware {
  const maxTokens = options.maxTokens ?? DEFAULT_MAX_TOKENS;
  const windowMs = options.windowMs ?? DEFAULT_WINDOW_MS;
  const maxRequestTokens = options.maxRequestTokens;
  const keyGenerator = options.keyGenerator ?? defaultKeyGenerator;
  const estimateTokens = options.estimateTokens ?? defaultEstimateTokens;
  const statusCode = options.statusCode ?? 429;
  const statusCodeOversize = options.statusCodeOversize ?? 413;
  const message = options.message ?? 'Token budget exceeded for this window.';
  const messageOversize =
    options.messageOversize ??
    'Request exceeds the per-request token limit.';
  const skip = options.skip;

  // Object.create(null) keeps __proto__/constructor keys from corrupting
  // the store if a buggy keyGenerator returns one of those names.
  const store = Object.create(null) as Record<string, Entry>;

  // Sweep expired buckets so a long-lived process doesn't grow unbounded.
  const cleanup = setInterval(() => {
    const now = Date.now();
    for (const k of Object.keys(store)) {
      if (store[k].resetTime < now) {
        delete store[k];
      }
    }
  }, windowMs);
  if (typeof cleanup.unref === 'function') cleanup.unref();

  const handler: RequestHandler = (req, res, next) => {
    if (skip?.(req)) {
      next();
      return;
    }

    let estimated = 0;
    try {
      estimated = estimateTokens(req);
    } catch {
      estimated = 0;
    }
    if (!Number.isFinite(estimated) || estimated < 0) estimated = 0;

    // Per-request cap (rejected before counting against the budget so a
    // single oversized request can't single-handedly drain a window).
    if (maxRequestTokens !== undefined && estimated > maxRequestTokens) {
      res.setHeader('X-Token-Budget-Limit', maxTokens.toString());
      res.setHeader('X-Token-Budget-Request-Cost', estimated.toString());
      res.status(statusCodeOversize).json({
        error: messageOversize,
        requestTokens: estimated,
        maxRequestTokens,
      });
      return;
    }

    const key = keyGenerator(req);
    const now = Date.now();
    let entry = store[key];
    if (!entry || entry.resetTime < now) {
      entry = { used: 0, resetTime: now + windowMs };
      store[key] = entry;
    }

    const projected = entry.used + estimated;
    const remaining = Math.max(0, maxTokens - entry.used);
    const resetSeconds = Math.ceil((entry.resetTime - now) / 1000);

    res.setHeader('X-Token-Budget-Limit', maxTokens.toString());
    res.setHeader('X-Token-Budget-Used', entry.used.toString());
    res.setHeader('X-Token-Budget-Remaining', remaining.toString());
    res.setHeader('X-Token-Budget-Reset', resetSeconds.toString());
    res.setHeader('X-Token-Budget-Request-Cost', estimated.toString());

    if (projected > maxTokens) {
      res.setHeader('Retry-After', resetSeconds.toString());
      res.status(statusCode).json({
        error: message,
        used: entry.used,
        maxTokens,
        retryAfter: resetSeconds,
      });
      return;
    }

    // Charge the budget and continue.
    entry.used = projected;
    next();
  };

  const middleware = handler as TokenBudgetMiddleware;
  middleware.close = () => clearInterval(cleanup);
  middleware.inspect = (key: string) => {
    const e = store[key];
    if (!e) return null;
    return { used: e.used, resetTime: e.resetTime };
  };
  return middleware;
}

export default tokenBudget;
